src/lib.rs

```diff
@@ -0,0 +1,253 @@
+//
+// Subject to the condition set forth below, permission is hereby granted to any person obtaining a
+// copy of this software, associated documentation and/or data (collectively the "Software"), free
+// of charge and under any and all copyright rights in the Software, and any and all patent rights
+// owned or freely licensable by each licensor hereunder covering either (i) the unmodified
+// Software as contributed to or provided by such licensor, or (ii) the Larger Works (as defined
+// below), to deal in both
+//
+// (a) the Software, and
+// (b) any piece of software and/or hardware listed in the lrgrwrks.txt file
+// if one is included with the Software (each a "Larger Work" to which the Software is contributed
+// by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create derivative works
+// of, display, perform, and distribute the Software and make, use, sell, offer for sale, import,
+// export, have made, and have sold the Software and the Larger Work(s), and to sublicense the
+// foregoing rights on either these or other terms.
+//
+// This license is subject to the following condition: The above copyright notice and either this
+// complete permission notice or at a minimum a reference to the UPL must be included in all copies
+// or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+//! An immutable cactus stuck (also called a spaghetti stack or parent pointer tree). A cactus
+//! stack is a (possibly empty) node with a (possibly null) pointer to a parent node. Pushing
+//! and popping return new nodes rather than mutating existing nodes.
+//!
+//! ```
+//! use cactus::Cactus;
+//! let c = Cactus::new();
+//! assert!(c.is_empty());
+//! let c2 = c.push(1);
+//! assert_eq!(c2.len(), 1);
+//! assert_eq!(*c2.val().unwrap(), 1);
+//! let c3 = c2.pop().unwrap();
+//! assert!(c3.is_empty());
+//! ```
+//!
+//! From a given node one can create multiple sub-stacks:
+//!
+//! ```
+//! use cactus::Cactus;
+//! let c = Cactus::new().push(1);
+//! let c2 = c.push(2);
+//! let c3 = c.push(3);
+//! assert!(c2 != c3);
+//! assert_eq!(c2.iter().cloned().collect::<Vec<_>>(), [2, 1]);
+//! assert_eq!(c3.iter().cloned().collect::<Vec<_>>(), [3, 1]);
+//! ```
+
+use std::fmt;
+use std::rc::Rc;
+
+/// An immutable cactus stack node. May be empty or contain a value; may have a pointer to a parent
+/// or not.
+#[derive(Clone, Default)]
+pub struct Cactus<T> {
+    node: Option<Rc<Node<T>>>
```

Comment:
I was trying to figure out why you use an `Rc` here.

It's because the values in the nodes can come from different incompatible lifetimes?

I suspect if you were to clone `val` in push, the `Rc` could go? Not sure, and I'm pretty sure you have tried lots of variations and found this to be the only one you can get rust to accept (?).